recursion.rb

```diff
@@ -1,19 +1,68 @@
 
 def fact(n)
+  if n == 1 || n == 0
+    return 1
+  elsif n < 0
+    raise ArgumentError.new("n must be greater than or equal to 0")
+  else
+    return n * fact(n-1)
+  end
 end
 
 def reverse(s)
+  var = ""
+  if s == ""
+    return var
+  else
+    var = s[-1] + reverse(s.chop)
+  end
+  return var
 end
 
-def nested(s)
+def bunny(n)
+  if n == 0
+    return 0
+  else
+    return 2 + bunny(n-1)
+  end
 end
 
-# Factorial Tests
-raise "factorial broke - fact(4)" unless fact(4) == 24
-raise "factorial broke - fact(0)" unless fact(0) == 1
-puts "passes all factorial tests"
-
-# Reverse Tests
+# This works with the examples given. However, in the case of something like "()(()())",
+# it fails. Technically, that example is nested correctly, but in a horrid manner. I
+# think the example we did during class, where ( = 1 and ) = -1 would be a better
+# approach, in that case, summing the numbers and returning false if the total became
+# negative.
+def nested(s)
+  result = true
+  if s.length % 2 == 1
+    result = false
+    return result
+  elsif s[0] == s[-1]
+    result = false
+    return result
+  elsif s.length >= 2
+    return nested(s[1..-2])
+  end
+  return result
+end
```

Comment:
Looks good, with a few simplifications for readability:

```
def nested(s)
   if s.length % 2 == 1
     return false
   elsif s[0] == s[-1]
     return false
   elsif s.length >= 2
     return nested(s[1..-2])
    else
     return true
   end
end
```

Basically if you can just return something, don't make a variable to store the information and then later return it.